File: problems/0002.add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        l1list = []
        l2list = []
        while l1 or l2:
            if l1:
                l1list.append(str(l1.val))
                l1 = l1.next
            if l2:
                l2list.append(str(l2.val))
                l2 = l2.next
        l1list.reverse()
        l2list.reverse()
        sum = list(str(int("".join(l1list)) + int("".join(l2list))))
        sum.reverse()
        node = temp = ListNode(int(sum[0]))
        for i in range(1, len(sum)):
            temp.next = ListNode(int(sum[i]))
            temp = temp.next
        return node

File: problems/0002.add-two-numbers/addTwoNumbers.js
/**
 * 2. Add Two Numbers
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

const addTwoNumbers = function (l1, l2) {
  let head = null,
    temp = null,
    carry = 0;
  while (l1 !== null || l2 !== null) {
    let sum = carry;
    if (l1 !== null) {
      sum += l1.val;
      l1 = l1.next;
    }
    if (l2 !== null) {
      sum += l2.val;
      l2 = l2.next;
    }

    let node = new ListNode(sum % 10);
    carry = Math.floor(sum / 10);

    if (temp === null) {
      temp = head = node;
    } else {
      temp.next = node;
      temp = temp.next;
    }
  }
  if (carry > 0) {
    temp.next = new ListNode(carry);
  }

  return head;
};

module.exports = addTwoNumbers;
